dbms/src/Functions/FunctionsJson.h

```diff
@@ -386,25 +432,59 @@ class FunctionCastJsonAsString : public IFunction
             {
-                size_t next_offset = offsets_from[i];
-                size_t json_length = next_offset - current_offset - 1;
-                if unlikely (isNullJsonBinary(json_length))
+                size_t current_offset = 0;
+                for (size_t i = 0; i < block.rows(); ++i)
                 {
-                    vec_null_map[i] = 1;
+                    size_t next_offset = offsets_from[i];
+                    size_t json_length = next_offset - current_offset - 1;
+                    if unlikely (isNullJsonBinary(json_length))
+                        vec_null_map[i] = 1;
+                    else
+                    {
+                        JsonBinary json_binary(
+                            data_from[current_offset],
+                            StringRef(&data_from[current_offset + 1], json_length - 1));
+                        json_binary.toStringInBuffer(write_buffer);
+                    }
+                    writeChar(0, write_buffer);
+                    offsets_to[i] = write_buffer.count();
+                    current_offset = next_offset;
                 }
-                else
+            }
+            else
+            {
+                ColumnString::Chars_t container_per_element;
+                size_t current_offset = 0;
+                for (size_t i = 0; i < block.rows(); ++i)
                 {
-                    JsonBinary json_binary(
-                        data_from[current_offset],
-                        StringRef(&data_from[current_offset + 1], json_length - 1));
-                    json_binary.toStringInBuffer(write_buffer);
+                    size_t next_offset = offsets_from[i];
+                    size_t json_length = next_offset - current_offset - 1;
+                    if unlikely (isNullJsonBinary(json_length))
+                        vec_null_map[i] = 1;
+                    else
+                    {
+                        JsonBinary::JsonBinaryWriteBuffer element_write_buffer(container_per_element);
+                        JsonBinary json_binary(
+                            data_from[current_offset],
+                            StringRef(&data_from[current_offset + 1], json_length - 1));
+                        json_binary.toStringInBuffer(element_write_buffer);
+                        size_t orig_length = element_write_buffer.count();
+                        auto byte_length = charLengthToByteLengthFromUTF8(
+                            reinterpret_cast<char *>(container_per_element.data()),
+                            orig_length,
+                            tidb_tp->flen());
+                        byte_length = std::min(byte_length, orig_length);
+                        if (byte_length < element_write_buffer.count())
+                            context.getDAGContext()->handleTruncateError("Data Too Long");
+                        write_buffer.write(reinterpret_cast<char *>(container_per_element.data()), byte_length);
```

Comment:
how about
```suggestion
                        auto start_pos = write_buffer.offset();
                        JsonBinary json_binary(
                            data_from[current_offset],
                            StringRef(&data_from[current_offset + 1], json_length - 1));
                        json_binary.toStringInBuffer(write_buffer);
                        auto end_pos = write_buffer.offset();
                        auto orig_length = end_pos - start_pos;
                        auto byte_length = charLengthToByteLengthFromUTF8(
                            reinterpret_cast<char *>(write_buffer.data() + start_offset),
                            orig_length,
                            tidb_tp->flen());
                        if (byte_length < orig_length)
                        {
                            context.getDAGContext()->handleTruncateError("Data Too Long");
                            write_buffer.setOffset(start_pos + byte_length);
                        }
```
?

Reply:
To avoid one more memcpy.

Reply:
Yeah，I just think this code path is not common used(because cast json as fixed length char is a little strange), and even if it is used the performance won't drop significantly, thus choose to use the temporary buffer here to make code more easier.